feature(api-v2): Generate client test data for ontologies, resources, search, and lists

benjamingeer opened the pull request:
This PR incorporates more API v2 routes into the client code generation framework so they can return client test data.

- [x] ontologies route
- [x] resources route
- [x] search route
- [x] lists route

Resolves #1548.

tobiasschweizer commented:
In v2 test data for resources, I cannot find JSON data for creation and metadata update ops.

tobiasschweizer commented:
see https://github.com/dasch-swiss/knora-api-js-lib/blob/9e16288916c2e2a5a66b7d0d9f4448c7dd911438/src/api/v2/resource/resources-endpoint.spec.ts#L96-L648 for the tests I have written so far (test data hard coded)

tobiasschweizer commented:
I think the the values themselves do not have to be tested thoroughly because this is already done in https://github.com/dasch-swiss/knora-api-js-lib/blob/wip/create-resource/src/api/v2/values/values-endpoint.spec.ts with test data generated by Knora.

benjamingeer commented:
> In v2 test data for resources, I cannot find JSON data for creation and metadata update ops.

I just included the stuff you asked for 😜:

https://github.com/dasch-swiss/knora-api/pull/1546#issuecomment-562212304

benjamingeer commented:
Will add test data for the update operations, too.

tobiasschweizer commented:
> Will add test data for the update operations, too.

Thanks!

tobiasschweizer commented:
In the meantime I am changing from static to generated v2 test data files. Works already for the ontologies! :-)

My goal is to get rid of static test data at all.

tobiasschweizer commented:
there is a failing test in  ExceptionHandlerR2RSpec: https://github.com/dasch-swiss/knora-api/pull/1549/checks?check_run_id=341570711#step:6:1310

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/routing/v2/ListsRouteV2.scala:
```diff
@@ -19,64 +19,114 @@
+    override val description: String = "An endpoint for working with Knora lists."
+    override val functions: Seq[ClientFunction] = Seq.empty
+
+    def knoraApiPath: Route = getList ~ getNode
```
Could you add a comment that this is actually the route definition?

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/routing/v2/OntologiesRouteV2.scala:
```diff
@@ -19,85 +19,270 @@
     private val LAST_MODIFICATION_DATE = "lastModificationDate"
 
-    def knoraApiPath: Route = {
+    def knoraApiPath: Route = dereferenceOntologyIri ~ getOntologyMetadata ~ updateOntologyMetadata ~ getOntologyMetadataForProjects ~
```
Could you add a comment that this is actually the route definition?

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala:
```diff
@@ -127,138 +147,61 @@ class ResourcesRouteV2(routeData: KnoraRouteData) extends KnoraRoute(routeData)
         }
     }
 
+    def knoraApiPath: Route = createResource ~ updateResourceMetadata ~ getResourcesInProject ~
```
Could you add a comment that this is actually the route definition?

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/routing/v2/SearchRouteV2.scala:
```diff
@@ -19,32 +19,49 @@
     private val OFFSET = "offset"
     private val LIMIT_TO_STANDOFF_CLASS = "limitToStandoffClass"
 
+    def knoraApiPath: Route = fullTextSearchCount ~ fullTextSearch ~ gravsearchCountGet ~ gravsearchCountPost ~
```
Could you add a comment that this is actually the route definition?

benjamingeer commented:
> there is a failing test in ExceptionHandlerR2RSpec

That's not a failing test, the test is supposed to throw an exception. The GitHub CI build is actually failing here, but I don't know why:

https://github.com/dasch-swiss/knora-api/pull/1549/checks?check_run_id=341570711#step:6:2012

benjamingeer commented:
> Could you add a comment that this is actually the route definition?

I've added it to all routes.

I added more test requests/responses for the resources route. Do you want update requests for the ontologies route too?

tobiasschweizer commented:
> I've added it to all routes.

Great, thanks.

> I added more test requests/responses for the resources route. Do you want update requests for the ontologies route too?

Yes, we will need that too. 

Could you also create JSON test data for a resource preview (response to a creation request)?

tobiasschweizer commented:
The preview JSON data has a higher priority to finish https://github.com/dasch-swiss/knora-api-js-lib/pull/115, so if you need more time for the ontologies and want to do this in a subsequent PR, that's also fine.

benjamingeer commented:
> if you need more time for the ontologies and want to do this in a subsequent PR

Yes, because I'd like to finish #1509 first.

tobiasschweizer commented:
When I access http://localhost:3333/clientapi/typescript I now get

{
"error": "akka.http.scaladsl.model.IllegalUriException: Illegal URI reference: Invalid input '/', expected userinfo-char, pct-encoded or '@' (line 1, column 36): http://0.0.0.0:3333resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing\n                                   ^"
}

tobiasschweizer commented:
```
"anything:hasDecimal" : {
    "@type" : "knora-api:DecimalValue",
    "knora-api:decimalValueAsDecimal" : {
      "@type" : "x  sd:decimal",
      "@value" : "100000000000000.000000000000001"
    }
  }
```

I think this is too much precision

tobiasschweizer commented:
Maybe I could look at https://github.com/royNiladri/js-big-decimal or some similar lib.

tobiasschweizer commented:
if possible, could you also generate data for 

- update a resource's metadata submitting the last modification date
- response to metadata update
- response to resource deletion

benjamingeer commented:
> "error": "akka.http.scaladsl.model.IllegalUriException

Sorry, stupid mistake.

> update a resource's metadata submitting the last modification date

Yes and it looks like there's a bug, the last modification date should be required.

tobiasschweizer commented:
When I try to generate client api code I now get:

{"error":"org.knora.webapi.ClientApiGenerationException: Failed to get test data: The requested resource could not be found."}

However, I can get http://localhost:3333/v2/resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala:
```diff
@@ -376,7 +376,7 @@ class ResourcesRouteV2(routeData: KnoraRouteData) extends KnoraRoute(routeData)
     private def getResourcesPreviewTestResponse: Future[SourceCodeFileContent] = {
         for {
-            responseStr <- doTestDataRequest(Get(s"${baseApiUrl}resourcespreview/${SharedTestDataADM.AThing.iriEncoded}"))
+            responseStr <- doTestDataRequest(Get(s"$baseApiUrl/resourcespreview/${SharedTestDataADM.AThing.iriEncoded}"))
```
you are still missing v2: http://0.0.0.0:3333/resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing

tobiasschweizer replied:
Shall I fix it?

benjamingeer replied:
I'm doing it now along with the other stuff you asked for.

tobiasschweizer replied:
Ok, thank you.

benjamingeer commented:
> org.knora.webapi.ClientApiGenerationException

Yet another stupid mistake. I think I need to hibernate for the winter.

Can I change the file extensions to `.jsonld` instead of `.json`, or would that be a pain?

tobiasschweizer commented:
> Can I change the file extensions to .jsonld instead of .json, or would that be a pain?

I experienced some problems earlier with that extension in the past when loading the files in a JS env. But we can give it a try in a separate PR. I am supposed to publish a new version of knora-api-js-lib today for @lrosenth, so I just want to finish resource creation.

benjamingeer commented:
OK please try it now.

tobiasschweizer commented:
Ok, that's all I need to finish https://github.com/dasch-swiss/knora-api-js-lib/pull/115, thanks!

I spotted a potential issue for the automatic integration tests in Knora we want to set up (#1554): some of the data is dynamic (such as creation / modification dates) and the tests will fail because the assertions are hard-coded (but they won't fail because of an actual problem we would have to fix).

benjamingeer commented:
> some of the data is dynamic (such as creation / modification dates)

I think I've fixed this now. Let me know if it's OK to merge.

tobiasschweizer commented:
> I think I've fixed this now. Let me know if it's OK to merge.

Works :-)

benjamingeer commented:
> Works :-)

Did you forget to press the "Approve" button, or is there something else I need to do?

benjamingeer commented:
Thanks for your patience!

tobiasschweizer commented:
> Thanks for your patience!

No worries!

tobiasschweizer commented:
Once this is merged, could you update the branches for the modification date and the time value in Knora. I then have to regenerate the client data from these branches to integrate them in the PR is knora-api-js-lib.
